components/MaskedTransition/src/MDCMaskedTransition.m

```diff
@@ -0,0 +1,286 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import "MDCMaskedTransition.h"
+
+#import <MotionAnimator/MotionAnimator.h>
+
+#import "MDCMaskedPresentationController.h"
+#import "MDCMaskedTransitionMotionForContext.h"
+#import "MDCMaskedTransitionMotionSpec.h"
+
+// Math utilities
+
+static CGPoint centerOfFrame(CGRect frame) {
+  return CGPointMake(CGRectGetMidX(frame), CGRectGetMidY(frame));
+}
+
+static CGPoint anchorPointFromPosition(CGPoint position, CGRect bounds) {
+  return CGPointMake(position.x / bounds.size.width, position.y / bounds.size.height);
+}
+
+static CGRect frameCenteredAround(CGPoint position, CGSize size) {
+  return CGRectMake(position.x - size.width / 2,
+                    position.y - size.height / 2,
+                    size.width,
+                    size.height);
+}
+
+static CGFloat lengthOfVector(CGVector vector) {
+  return (CGFloat)sqrt(vector.dx * vector.dx + vector.dy * vector.dy);
+}
+
+@interface MDCMaskedTransition () <MDMTransitionWithPresentation, MDMTransitionWithFallback>
+@end
+
+@implementation MDCMaskedTransition {
+  UIView *_sourceView;
+  MDCMaskedPresentationController *_presentationController;
+  BOOL _shouldSlideWhenCollapsed;
+}
+
+- (instancetype)initWithSourceView:(UIView *)sourceView {
+  self = [super init];
+  if (self) {
+    _sourceView = sourceView;
+  }
+  return self;
+}
+
+- (id<MDMTransition>)fallbackTransitionWithContext:(id<MDMTransitionContext>)context {
+  return _shouldSlideWhenCollapsed ? nil : self;
+}
+
+#pragma mark - MDMTransitionWithPresentation
+
+- (UIModalPresentationStyle)defaultModalPresentationStyle {
+  return UIModalPresentationCustom;
+}
+
+- (UIPresentationController *)presentationControllerForPresentedViewController:(UIViewController *)presented
+                                                      presentingViewController:(UIViewController *)presenting
+                                                          sourceViewController:(UIViewController *)source {
+  _presentationController =
+      [[MDCMaskedPresentationController alloc] initWithPresentedViewController:presented
+                                                      presentingViewController:presenting
+                                                 calculateFrameOfPresentedView:_calculateFrameOfPresentedView];
+  return _presentationController;
+}
+
+- (void)startWithContext:(NSObject<MDMTransitionContext> *)context {
+  MDCMaskedTransitionMotionSpec spec = motionForContext(context);
+  if (context.direction == MDMTransitionDirectionForward) {
+    _shouldSlideWhenCollapsed = spec.shouldSlideWhenCollapsed;
+  }
+
+  MDMMotionAnimator *animator = [[MDMMotionAnimator alloc] init];
+  animator.shouldReverseValues = context.direction == MDMTransitionDirectionBackward;
+
+  // # Caching original state
```

Comment:
Is the hash mark doing anything here? In general, full sentences with punctuation, please.